Detectors/TOF/simulation/src/Detector.cxx

```diff
@@ -39,12 +39,28 @@ Detector::Detector(const char* Name, Bool_t Active)
     mTOFSectors[i] = 1;
 }
 
+Detector::Detector(const Detector& rhs)
+  : o2::Base::Detector(rhs),
+    mEventNr(0),
+    mTOFHoles(rhs.mTOFHoles),
+    mHitCollection(new TClonesArray("o2::tof::HitType")),
+    mMCTrackBranchId(-1)
+{
+  for (Int_t i = 0; i < Geo::NSECTORS; i++)
+    mTOFSectors[i] = rhs.mTOFSectors[i];
+}
+
+FairModule* Detector::CloneModule() const
+{
+  return new Detector(*this);
+}
+
 void Detector::Initialize() { o2::Base::Detector::Initialize(); }
 Bool_t Detector::ProcessHits(FairVolume* v)
 {
-  static auto* refMC = TVirtualMC::GetMC();
+  TMCThreadLocalStatic auto* refMC = TVirtualMC::GetMC();
 
-  static TLorentzVector position2;
+  TMCThreadLocalStatic TLorentzVector position2;
```

Comment:
I think we should now try to avoid static thread local variables altogether (they are really bad for performance). This can be done by making these variables a member of the Detector class.

Reply:
Ok;I will change such local variables specific to concrete detectors in dat members.
And what about these (see eg. TOF Detector::ProcessHits, but I think it is used by all detectors):
  static thread_local auto* refMC = TVirtualMC::GetMC();
Do you have something against using gMC, which is provided in TVirtualMC? See the implementation in root/montecarlo/vmc/inc/TVirtualMC.h
Or, if yes, we could then cache the pointer in a protected data member of Base::Detector class in Initialize(); but then we have a danger of using 0 pointer, as it cannot be initialized earlier, as it would point at the master object, while in ProcessHits() we need the object on worker. Using gMC guarantees that the user picks the right object.

Reply:
I started to modify static thread local variables in detectors, but I think it will be better that I just add thread_local declarations and we ask detector developers to change these in data members of the class. As according to the O2 coding guidelines  "static local variable s" should be avoided even when they are not declared thread local.

Reply:
I would actually prefer avoiding thread_local variables at all cost. Using them in critical code such as stepping functions leads to visible performance penalties (as seen by studies in AliRoot last year). This is why gMC should not used in favour of a local cache. Same is true for other static members.

Reply:
I am not proposing to keep thread_local static variables in the code for ever, but to put that until detectors will change the code. The existing static local variables are not compatible with MT mode, so the cannot stay as they are. I am just proposing an interim step so that we can have the stuff working with MT and then let detectors modify their code.